core/src/main/java/com/linecorp/armeria/internal/common/DefaultRequestTarget.java

```diff
@@ -0,0 +1,966 @@
+            switch (reqTarget.charAt(i)) {
+                case '/':
+                case '?':
+                case '#':
+                    return i;
+            }
+        }
+
+        return -1;
+    }
+
+    @Nullable
+    private static RequestTarget slowForClient(String reqTarget,
+                                               @Nullable URI schemeAndAuthority,
+                                               int pathPos) {
+        final Bytes fragment;
+        final Bytes path;
+        final Bytes query;
+        // Find where a query string and a fragment starts.
+        final int queryPos;
+        final int fragmentPos;
+        // Note: We don't start from `pathPos + 1` but from `pathPos` just in case path is empty.
+        final int maybeQueryPos = reqTarget.indexOf('?', pathPos);
+        final int maybeFragmentPos = reqTarget.indexOf('#', pathPos);
+        if (maybeQueryPos >= 0) {
+            // Found '?'.
+            if (maybeFragmentPos >= 0) {
+                // Found '#', too.
+                fragmentPos = maybeFragmentPos;
+                if (maybeQueryPos < maybeFragmentPos) {
+                    // '#' appeared after '?', e.g. ?foo#bar
+                    queryPos = maybeQueryPos;
+                } else {
+                    // '#' appeared before '?', e.g. #foo?bar.
+                    // It means the '?' we found is not a part of query string.
+                    queryPos = -1;
+                }
+            } else {
+                // No '#' in reqTarget.
+                queryPos = maybeQueryPos;
+                fragmentPos = -1;
+            }
+        } else {
+            // No '?'.
+            queryPos = -1;
+            fragmentPos = maybeFragmentPos;
+        }
+
+        // Split into path, query and fragment.
+        if (queryPos >= 0) {
+            if ((path = decodePercentsAndEncodeToUtf8(
+                    reqTarget, pathPos, queryPos,
+                    ComponentType.CLIENT_PATH, SLASH_BYTES)) == null) {
+                return null;
+            }
+
+            if (fragmentPos >= 0) {
+                // path?query#fragment
+                if ((query = decodePercentsAndEncodeToUtf8(
+                        reqTarget, queryPos + 1, fragmentPos,
+                        ComponentType.QUERY, EMPTY_BYTES)) == null) {
+                    return null;
+                }
+                if ((fragment = decodePercentsAndEncodeToUtf8(
+                        reqTarget, fragmentPos + 1, reqTarget.length(),
+                        ComponentType.FRAGMENT, EMPTY_BYTES)) == null) {
+                    return null;
+                }
+            } else {
+                // path?query
+                if ((query = decodePercentsAndEncodeToUtf8(
+                        reqTarget, queryPos + 1, reqTarget.length(),
+                        ComponentType.QUERY, SLASH_BYTES)) == null) {
+                    return null;
+                }
```

Comment:
If a query is empty (`?`), it is converted to `/`.
```java
final RequestTarget res = RequestTarget.forClient("/foo/bar?");
// `res.query()` returns `/` and the assertion fails.
assertThat(res.query()).isEqualTo("");
```
Should we return `EMPTY_BYTES` if it is empty?
```suggestion
                if ((query = decodePercentsAndEncodeToUtf8(
                        reqTarget, queryPos + 1, reqTarget.length(),
                        ComponentType.QUERY, EMPTY_BYTES)) == null) {
                    return null;
                }
```

Reply:
Oops! Let me fix this and add a test.